### PR TITLE
Improve image size and layers

### PR DESCRIPTION
**Pull request by williamdes**

### CODE REVIEW

**Comment by jbergstroem** [comment]
I started out with inlining all of `clean.sh` in the Dockerfile, but it was very tedious to work with as well as allowing me to lint directly via `shellcheck`.

How much space does this save?

**Comment by jbergstroem** [comment]
FYI! This introduces two warnings via hadolint: https://github.com/jbergstroem/mariadb-alpine/pull/75/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R18

**Comment by williamdes** [comment]
> I started out with inlining all of `clean.sh` in the Dockerfile, but it was very tedious to work with as well as allowing me to lint directly via `shellcheck`.
> 
> How much space does this save?

The file space (799 B) because even if you add it you may remove it but it still counts as a file

> shellcheck

Oh snap, that's a downside

I would argue that have everything contained into one Dockerfile is better for audits

File: Dockerfile [hunk 1]
```diff
@@ -13,19 +15,45 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
...
+  done && \
+  \
   touch /usr/share/mariadb/mysql_test_db.sql && \
   # allow anyone to connect by default
   sed -i -e 's/127.0.0.1/%/' /usr/share/mariadb/mysql_system_tables_data.sql && \
   mkdir /run/mysqld && \
   chown mysql:mysql /etc/my.cnf.d/ /run/mysqld /usr/share/mariadb/mysql_system_tables_data.sql
 
+COPY sh/run.sh /
+# The one installed by MariaDB was removed in the clean step above
```
Comment (jbergstroem) [inline Dockerfile:54]: FWIW: The reason I remove `removeip` is that it is a multi-megabyte binary that fills little to no purpose within docker. This comment could mention this so a newcomer to the project doesn't think this is something that needs to change.

File: Dockerfile [hunk 2]
```diff
@@ -3,6 +3,8 @@ FROM alpine:3.12
 # https://github.com/opencontainers/image-spec/blob/master/annotations.md
 ARG BUILD_DATE
 ARG VCS_REF
+ARG MARIADB_VERSION="10.4.18-r0"
```
Comment (jbergstroem) [inline Dockerfile:6]: Can you skip this for now? I will do bigger changes here to push multiple versions.

**Review by jbergstroem — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -13,19 +15,45 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
...
 
-COPY sh/clean.sh /clean.sh
-
-RUN apk add --no-cache mariadb=10.4.18-r0 && \
-  /bin/sh /clean.sh && \
-  # removed in cleaning
+RUN \
+  apk add --no-cache mariadb=${MARIADB_VERSION} && \
+  \
+  # Clean the image
```
Comment (jbergstroem) [inline Dockerfile:21]:
```suggestion
```

File: Dockerfile [hunk 4]
```diff
@@ -13,19 +15,45 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
...
+    usr/share/mariadb/mysql_system_tables.sql$ \
+    usr/share/mariadb/mysql_performance_tables.sql$ \
+    usr/share/mariadb/mysql_system_tables_data.sql$ \
+    usr/share/mariadb/maria_add_gis_sp_bootstrap.sql$ \
+    usr/share/mariadb/fill_help_tables.sql$" | \
+    tr -d " \t\n\r" | sed -e 's/usr/|usr/g' -e 's/^.//') && \
+  INSTALLED_FILES="$(apk info -q -L mariadb-client | tail -n +2) \
+  $(apk info -q -L mariadb-common | tail -n +2) \
+  $(apk info -q -L mariadb | tail -n +2)" && \
+  \
```
Comment (jbergstroem) [inline Dockerfile:42]:
```suggestion
```

File: Dockerfile [hunk 5]
```diff
@@ -13,19 +15,45 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
...
+  for path in $(echo "${INSTALLED_FILES}" | grep -v -E "${TO_KEEP}"); do \
+    eval rm -rf "${path}"; \
+  done && \
+  \
```
Comment (jbergstroem) [inline Dockerfile:46]:
```suggestion
```

File: Dockerfile (hunk 1 above)
Comment (williamdes) [reply Dockerfile:54]: Yeah, Indeed a 3MB file 😄 
I will edit my comment

**Review by jbergstroem — COMMENTED** [review]
Nit: no need for extra spaces

**Comment by jbergstroem** [comment]
> > How much space does this save?
> 
> The file space (799 B) because even if you add it you may remove it but it still counts as a file

Yeah, so, I had a space vs convenience battle in my head wrt this a while back and felt that when I start optimizing in the byte range I might be taking it too far.

> 
> > shellcheck
> 
> Oh snap, that's a downside

It should be invoked indirectly via the hadolint tho (as it did via CI)

> 
> I would argue that have everything contained into one Dockerfile is better for audits

Yeah, I can buy that argument.

File: Dockerfile (hunk 4 above)
Comment (williamdes) [reply Dockerfile:38]: Are you sure to remove them ?
It adds some spacing

**Comment by williamdes** [comment]
All should be solved now :)

**Comment by jbergstroem** [comment]
@williamdes can you rebase on latest master pls?

**Comment by williamdes** [comment]
> @williamdes can you rebase on latest master pls?

Done :)

File: Dockerfile [hunk 6]
```diff
@@ -13,19 +14,42 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
...
+  for path in $(echo "${INSTALLED_FILES}" | grep -v -E "${TO_KEEP}"); do \
+    eval rm -rf "${path}"; \
+  done && \
   touch /usr/share/mariadb/mysql_test_db.sql && \
   # allow anyone to connect by default
   sed -i -e 's/127.0.0.1/%/' /usr/share/mariadb/mysql_system_tables_data.sql && \
   mkdir /run/mysqld && \
   chown mysql:mysql /etc/my.cnf.d/ /run/mysqld /usr/share/mariadb/mysql_system_tables_data.sql
 
+# The one installed by MariaDB was removed in the clean step above, the size was to significant
```
Comment (jbergstroem) [inline Dockerfile:49]:
```suggestion
# The one installed by MariaDB was removed in the clean step above due to its large footprint
```

File: Dockerfile [hunk 7]
```diff
@@ -13,19 +14,42 @@ LABEL org.opencontainers.image.created=$BUILD_DATE \
   org.opencontainers.image.schema-version="1.0.0-rc.1" \
   org.opencontainers.image.license="MIT"
 
-COPY sh/clean.sh /clean.sh
+SHELL ["/bin/ash", "-euo", "pipefail", "-c"]
```
[the remark this answers is not in the document]
Comment (williamdes) [reply Dockerfile:17]: Yeah, copied from the hadolint github docs for the error

File: Dockerfile (hunk 6 above)
Comment (jbergstroem) [reply Dockerfile:49]: Can you incorporate this to your commit? Not sure how I can do that from my end. Will land after!

**Comment by williamdes** [comment]
All good, ready to merge ;p

**Review by jbergstroem — APPROVED** [review]

**Comment by jbergstroem** [comment]
> All good, ready to merge ;p

Thanks for helping out!

**Comment by williamdes** [comment]
> > All good, ready to merge ;p
> 
> Thanks for helping out!

Let me know if you need help

**Pull request merged**